Use Reverse Z for the depth buffer

Khasehemwy opened the pull request:
Add support for reverse-z depth buffer resolve godotengine/godot-proposals#3539
@BastiaanOlij    @clayjohn
  
**These changes break shader compatibility!**

I finished part of it.

**Have finished:**
1. operations related to `Projection` class (set_depth_correction() function).
2. most depth clear values in cpp files are inverted (e.g. 1.0 changed to 0.0) (canvas not use reversed-z).
3. most depth test function inverted (e.g. LESS changed to GREATER) (canvas not use reversed-z).
4. glsl for skybox.
5. light shadow. 
        (~directional shadow seems to be some problem~, mentioned in the conversation below)
        (update: [clayjohn](https://github.com/clayjohn) fixed directional shadow problem)
6. FSR2 (mentioned in the conversation below)
7. SSR, SSAO, SSIL, Dof, SDFGI, VoxelGI, VolumetricFog (Thanks [clayjohn](https://github.com/clayjohn) for finishing them)
8. XR rendering (Thanks [BastiaanOlij](https://github.com/BastiaanOlij) for finishing them)

**Unfinished and known issues:**
1. ~some problems with the xr part when using OpenGL API (mentioned in the conversation below)~
2. ~SSAO is incorrect (most post processing effects will have problems)~

Khasehemwy committed:
Use Reverse Z for the depth buffer

AThousandShips commented on core/math/projection.cpp:
These changes break compatibility, these should not be changed without some options

Khasehemwy replied:
Yes, the result of our discussion in the issue was to break compatibility

AThousandShips replied:
Oh got it (please mention in the OP, it's pretty important)

AThousandShips replied:
Though can you point me to where the specific changes to `Projection` was agreed on? I don't see that specifically mentioned

Khasehemwy replied:
> Though can you point me to where the specific changes to `Projection` was agreed on? I don't see that specifically mentioned

All changes in projection.cpp file

AThousandShips replied:
Yes? I'm asking where in the proposal this was agreed on

Khasehemwy replied:
Sorry, I misunderstood. There is no specific mention of modifications to `Projection` in the proposal. The result of the discussion is without compatibility code. I think it is most appropriate to modify it here.

AThousandShips replied:
Then I don't think that's necessarily what was agreed upon by the team, this changes code using this, it's exposed to scripting (and a lot of places) so if it wasn't discussed that these changes were part of it I don't think it was clearly agreed on

Khasehemwy replied:
These modifications ensure that the return value will not change after using reversed-z. Maybe this will not break the compatibility of the exposed functions?

AThousandShips replied:
Does it give the same results with `get_z_near/far`?

Khasehemwy replied:
> Does it give the same results with `get_z_near/far`?

Yes

AThousandShips replied:
Good, I did ask that though so would have been good to have that clarified to begin with :)

AThousandShips replied:
Also does:
* `get_projection_plane`
* `get_viewport_half_extents`
* `get_far_plane_half_extents`

Also remain the same?

AThousandShips replied:
It's not directly exposed, though it is used for static methods so needs to be changed to work correctly with the bindings, the `create_` methods

Khasehemwy replied:
All right, I think the compatibility of `create_` methods will not be maintained by my code.

AThousandShips reviewed (dismissed):
Some of these are exposed, and unless they are unchanged by this they break compatibility

Calinou reviewed (commented):
Tested locally, it mostly works as expected.

**Testing project:** https://github.com/godotengine/godot-proposals/files/10502399/test_zbuffer_precision.zip

Directional shadows for large objects break, even if the object is subdivided (each plane here is subdivided into 21×21 parts of equal size). I've also tried adjusting **Pancake Size** in DirectionalLight3D to no avail.

*View videos in fullscreen to make the difference more noticeable. Focus on the area between the orange and white planes.*

### Forward+

#### Before

https://github.com/godotengine/godot/assets/180032/d7b2fddb-ba6d-4728-b329-6c59b04c2626

#### After *(this PR)*

*Notice the phantom shadow appearing at the beginning of the video, which isn't visible in `master`.*

https://github.com/godotengine/godot/assets/180032/973939df-36cf-4749-8edb-35f8a6276831

### Mobile

*The shadow issues on the cyan spheres also appear in `master` without this PR.*

https://github.com/godotengine/godot/assets/180032/ec19a7e4-b39e-4d1f-9c5f-e50f61386df3

### Compatibility

*This PR doesn't seem to affect precision in Compatibility so far.*

https://github.com/godotengine/godot/assets/180032/39e6b936-fbec-4a35-8920-ae4acc8caba3

BastiaanOlij commented:
This needs more testing by myself and clay and there are a few change we likely want. But it is important to know that there is consensus within the rendering team that the compatibility breakage is acceptable as this will effect really limited edge cases.
Anyone who does unprojection in shaders is not effected as the unprojection deals with this. Only when working with the z values as is could result in breakage. The benefits far outweigh the cons here.
We do have a few suggestions for limiting breakage, those will be forthcoming soon.

BastiaanOlij commented:
@Khasehemwy one important change you might be able to do, we don't want the Z-flip to be part of the code for populating the the projection matrix. So we leave all the setters etc. alone.

Instead the Projection class needs a single `reverse_z` function that applies the flip. This way we can keep using the projection matrix as is, we can keep the culling logic as is, we simply call the `reverse_z` early on in the `render_scene` logic, maybe as part of `render_scene_data` so only the rendering happens with the matrix with Z reversed.

That has far less impact than doing this early on.

Khasehemwy commented:
> @Khasehemwy one important change you might be able to do, we don't want the Z-flip to be part of the code for populating the the projection matrix, but only apply this when populating the GPU buffers. So we leave all the setters etc. alone.
> 
> The Projection class needs a single `reverse_z` function that applies the flip. This way we can keep using the projection matrix as is, we can keep the culling logic as is, we simply call the `reverse_z` early on in the `render_scene` logic, maybe as part of `render_scene_data` so only the rendering happens with the matrix with Z reversed.
> 
> That has far less impact than doing this early on.

I changed the projection matrix back to normal z, then added reverse z-related operations, and added reverse-related operations where I thought they needed to be added.

In addition, I found that there is a problem with the Dual Paraboloid shadow mode of Omni Light. It seems that the official version also has problems. It may not be caused by my changes.

Khasehemwy commented on core/math/projection.cpp:
```diff
@@ -267,11 +289,12 @@ void Projection::set_perspective(real_t p_fovy_degrees, real_t p_aspect, real_t
 
 	set_identity();
 
+	// Using reversed-z
```
It should be deleted later

Khasehemwy commented on core/math/projection.cpp:
```diff
@@ -808,6 +884,10 @@ bool Projection::is_orthogonal() const {
 }
 
+bool Projection::is_reversed_z() const {
+	return columns[2][2] > 0;
```
The situation of far < near is not considered

BastiaanOlij replied:
I don't think so either, I think we protect that on `Camera3D` anyway but if not, it's an acceptable assumption.

roalyr commented:
Will this not be an issue (asking just in case): https://github.com/roalyr/godot-for-3d-open-worlds/issues/15 ?

Ansraer commented:
Ok, we talked about this in the rendering meeting today.

We would prefer it if all the `if(is_reversed_depth)` code was removed; once the corrections have been applied those functions shouldn't be called anymore anyways.
In addition, it would be nice if there was the option (controlled by a boolean) to perform rev z in `set_depth_correction` and `create_depth_correction`, defaulting to `true`.

Once those changes have been made, there shouldn't be anything blocking this PR from being merged.
Great Work!

DarioSamo commented:
FSR2 will likely need some changes.

https://github.com/godotengine/godot/blob/b8f106acb24387d2f2c746a586edc0728fd10aaf/servers/rendering/renderer_rd/effects/fsr2.cpp#L523-L533

`"\n#define FFX_FSR2_OPTION_INVERTED_DEPTH 1\n"` must be added to the define (after `"\n#define FFX_FSR2_OPTION_HDR_COLOR_INPUT 1\n"` seems like a good spot).

https://github.com/godotengine/godot/blob/b8f106acb24387d2f2c746a586edc0728fd10aaf/servers/rendering/renderer_rd/effects/fsr2.cpp#L811

Should be changed to:

```context->fsr_desc.flags = FFX_FSR2_ENABLE_HIGH_DYNAMIC_RANGE | FFX_FSR2_ENABLE_DEPTH_INVERTED;```

I haven't tested these changes yet or if they're necessary but at least according to the FSR2 documentation they'd be required.

Khasehemwy commented:
> We would prefer it if all the `if(is_reversed_depth)` code was removed; once the corrections have been applied those functions shouldn't be called anymore anyways.

But, the implementation of functions such as `get_z_far` is different in reversed-z and normal-z. If the `projection matrix` is not directly modified like the beginning, I think we can only add a judgment on whether it is reversed-z or not in `get_z_far`.

roalyr commented:
Why must far and near planes be reconstructed from matrix and normalized if they can simply be declared as `Plane(0.0, 0.0, -1.0, z_far_val)`, for instance? Besides, I've mentioned before that large z-far values mess up plane normalization and you get an empty plane.

Khasehemwy commented:
It seems to be the same problem as @Calinou mentioned before. I tried the initial version of PR and had problems too.

I haven't figured out what causes it yet. There are definitely a lot of things that need to be changed. I will change the basic rendering process, and other things will probably need to be changed by the official later.😢

Khasehemwy commented:
Oh, I agree to change the relevant operations to set_depth_correction(). My focus is actually here:
> We would prefer it if all the if(is_reversed_depth) code was **removed**; once the corrections have been applied those functions shouldn't be called anymore anyways.
>> the implementation of functions such as `get_z_far` is different in reversed-z and normal-z.
>> **I think we can only add a judgment on whether it is reversed-z or not, in `get_z_far`**

The relevant code is:
```cpp
real_t Projection::get_z_far() const {
	const real_t *matrix = (const real_t *)columns;
	Plane new_plane;
	if (is_reversed_z()) {
		new_plane = Plane(matrix[3] + matrix[2],
				matrix[7] + matrix[6],
				matrix[11] + matrix[10],
				matrix[15] + matrix[14]);
	} else {
		new_plane = Plane(matrix[3] - matrix[2],
				matrix[7] - matrix[6],
				matrix[11] - matrix[10],
				matrix[15] - matrix[14]);
	}
...
```

DarioSamo commented:
> The relevant code is:

Am I wrong in assuming there would be no need for this if you just didn't modify the matrix at all from its current representation already and made the reverse Z a separate correction instead? It sounds to me like the near and far planes should be identical regardless of whether reverse Z is used or not, it just looks to me like you need the branching because you're modifying the matrix itself instead of just applying a correction before rendering.

Is that correct? Or does using reverse Z actually imply major differences as to what the near and far values mean?

Khasehemwy commented:
I understand, apply reverse when finally passing it to the GPU, so that I don't have to consider the reversed-z situation before.

But I see that both `create_depth_correction()` and `get_z_far()` seem to be exposed to the user. If the user uses `create_depth_correction()`, gives the result to the projection matrix, and then uses `get_z_far()`, it seems that the wrong value will be obtained.

DarioSamo commented:
> I understand, apply reverse when finally passing it to the GPU, so that I don't have to consider the reversed-z situation before.
> 
> But I see that both `create_depth_correction()` and `get_z_far()` seem to be exposed to the user. If the user uses `create_depth_correction()`, gives the result to the projection matrix, and then uses `get_z_far()`, it seems that the wrong value will be obtained.

`create_depth_correction` doesn't actually correct the current matrix AFAIK, it just sets the matrix to a fixed set of values to make a correction matrix.

https://github.com/godotengine/godot/blob/c05506f042fbaae40971b3319e262df983af80ce/core/math/projection.cpp#L796-L815

Therefore you'll never get a valid Z Far value out of this, as it's a matrix intended purely for being used as (correction * projection) in operations later. I don't think you need to worry about all methods working correctly once operations like these have been performed: hence why it's probably better to just not worry about reversing the matrix itself and leaving it as is and give that responsibility to a correction matrix instead.

Khasehemwy commented:
After OpenGL uses reversed-z here, it is still [-1, 1], but the far distance is closer to -1, which should have no effect on the accuracy. 

What we discussed in the issue before is that all APIs use reversed-z method. so that there is no need to deal with two situations, which facilitates subsequent development.

novaplusplus commented:
> I have another question:
> 
> I see the PR adds reverse Z to the GLES3 renderer without the ability turn it off.
> 
> However, in most cases we must **not** use reverse Z with GLES3 because you need GL_ARB_clip_control which is often not available with GLES3. In fact I don't see `glClipControl` being called at all.
> 
> Reverse Z gives LESS precision (not more) when not using `GL_ARB_clip_control / glClipControl`.

I'll start by saying that low-level graphics programming is definitely not my wheelhouse, so this might not be of any use, but I did learn about an alternate method to potentially do reverse-Z in OpenGL, involving using `glDepthRange`. This old Outerra post talks about it: https://outerra.blogspot.com/2012/11/maximizing-depth-buffer-range-and.html (DirectX vs OpenGL section). And apparently this function is still available with GLES (if floating point depth buffers are available)

DarioSamo commented:
> I moved all reverse-related operations to `set_depth_correction()`. I also added `bool p_remap_z` and `bool p_restore` parameters. Because set_depth_correction() is used in OpenGL, p_remap_z is necessary, and during the process of adapting shadows, I found that it would be more convenient if I could control the projection matrix restoration, so p_restore was added.

Cool, thanks!

> I also looked at https://outera.blogspot.com/2012/11/maximizing-depth-buffer-range-and.html. I feel that OpenGL does not need this kind of processing, because for compatibility, it usually uses a 24-bit depth buffer. I tried a 24-bit depth buffer before, the effect of reversed-z was not good enough too.

It indeed doesn't do much, but Clay pointed out to me we'd be switching to reverse Z in OpenGL anyway for the sake of consistency of the shader behavior between renderers even if it doesn't provide much of a benefit. Otherwise you'd have to keep in mind that the range of values you get in one backend would be reversed in the other one.

Ansraer commented:
Ok, we looked at this PR again during this week's rendering meeting. Thanks for incorporating the changes we suggested last time, this is already looking way better.

There are a couple minor nitpicks we discovered and would like to see fixed:
1. `set_depth_correction` Thanks for moving all the rev_z stuff on the cpu side into this function, this is far better to reverse z early on and then having to be careful with anything on the cpu side that interacts with projection. However, it has become a bit complicated to understand what exactly it is `set_depth_correction` is actually doing, especially with the rather convoluted if statements. Some documentation would be really appreciated, especially as to what the purpose of `restore` is.
2. A couple of times you are doing something along the lines of `1.0 - depth` in the shader code, which is less than ideal since it results in an immediate loss of the additional precision. This should be avoided at all cost, the shader should **always** operate with rev z.
3. The light calculation in the shader apparently reverses the light depth only moments before sampling, which means that we have already lost any additional precision. Instead, the light matrix should be reversed before it is used to calculate the depth in the first place (which it might already be? reversing shadow depth one time too often could have caused the shadow artifacts @mrjustaguy has reported). Note that this change would mean that any and all modifications that happen to the shadow depth (such as e.g. shadow bias) now also need to be reversed.
4. It would be neat if you could get FSR2 working with this PR, if Dario is correct it might be as simple as adding two lines. If that doesn't work and you don't want to investigate it yourself, that's fine too, we could fix this in a follow-up pr.

If you have any questions, feel free to reach out to us on the [rendering channel on RC](https://chat.godotengine.org/channel/rendering).
___
@BastiaanOlij Could you maybe look at how this PR performs in XR? Sorry for pinging you directly, but I lack the permissions to request reviews from teams.

dsnopek commented on drivers/gles3/shaders/sky.glsl:
```diff
@@ -28,7 +28,7 @@ void main() {
 	uv_interp = vec2(vertex_attrib.x, vertex_attrib.y * -1.0);
 #endif
-	gl_Position = vec4(uv_interp, 1.0, 1.0);
+	gl_Position = vec4(uv_interp, -1.0, 1.0);
```
Perhaps this change should only happen within the `#ifdef USE_INVERTED_Y`?

dsnopek replied:
Hm, no, this isn't the issue. I tried reverting the change, and it messed up depth - I see the things behind rather than in front.

clayjohn commented on drivers/gles3/rasterizer_scene_gles3.cpp:
```diff
@@ -820,6 +820,9 @@ void RasterizerSceneGLES3::_draw_sky(RID p_env, const Projection &p_projection,
 		float aspect = p_projection.get_aspect();
 
 		camera.set_perspective(environment_get_sky_custom_fov(p_env), aspect, near_plane, far_plane);
+		Projection correction;
+		correction.set_depth_correction(false, true, false);
+		camera = correction * camera;
```
@dsnopek I think that this code might fix your issue. It needs to be moved outside the conditional so it applies regardless of whether sky custom fov is used. Can you test again, but with these lines moved below the conditional?

dsnopek replied:
No, unfortunately, that doesn't appear to fix it :-/

clayjohn commented:
> A couple of times you are doing something along the lines of 1.0 - depth in the shader code, which is less than ideal since it results in an immediate loss of the additional precision. This should be avoided at all cost, the shader should always operate with rev z.
The light calculation in the shader apparently reverses the light depth only moments before sampling, which means that we have already lost any additional precision. Instead, the light matrix should be reversed before it is used to calculate the depth in the first place (which it might already be? reversing shadow depth one time too often could have caused the shadow artifacts @mrjustaguy has reported). Note that this change would mean that any and all modifications that happen to the shadow depth (such as e.g. shadow bias) now also need to be reversed

I started looking into resolving this issue. Unfortunately, it is quite complex. Cubemap shadows are rendered using non-reverse-z, then they are linearized and scaled to match the dual paraboloid shadows and flipped to reverse-z. Dual paraboloid shadows are rendered with reverse-z. 

Omnilight shadows are not calculated in NDC space though, they are in a special space that is essentially the local light space scaled by the light range. Ultimately, what we calculate is the distance to the light in this modified light space. 

So right now, to make this work with reverse-z we are forced to flip the scaled light distance right before the shadow map comparison. 

This can't easily be fixed for a couple of reasons:
1. The dual paraboloid mapping relies on its on mapping instead of using perspective projection (so we can't get the benefits of reverse-z anyway)
2. All shadows share a sampler that is tuned for reverse-z. So at sample time, we need to be in reverse-z
3. Ultimately, the sample position is less important than the sample that is actually stored in the texture
4. Typically shadows are 16 bit only and don't benefit from reverse-z anyway.

I'm going to continue looking at this tomorrow as I suspect there is a way to implement this a bit cleaner now that I understand the problem. We might even be able to create a sudo-projection matrix that works.

Khasehemwy commented:
I am also looking at Cubemap shadows. It is no problem to directly use the reversed-z projection to generate the Shadow Map. But when rendering objects, it is a bit tricky to sample the Shadow Map. I can only use 1.0 - depth.

clayjohn commented:
I have a fix locally for the directional light issue. I will make a PR to your branch soon!

clayjohn commented:
@Khasehemwy I made a PR to your branch. If you merge the PR to your branch it will show up here as well https://github.com/Khasehemwy/godot/pull/1

clayjohn commented:
So far I have tested (on all backends):
1. regular depth precision 
2. shadows

With my PR both depth and shadows are working correctly. Including fixing the issue mrjustaguy reported. 

I have not tested post processing or GI which are other areas where things might break. Before this is merged we need to:
1. Merge my PR into Khasehemwy's branch
2. Test post processing and GI and confirm they continue to work

I think we should try to fix the WebXR regression after merging as it might take a bit for us to narrow down the issue.

violetfrost commented:
How does Reverse-Z impact user shaders? (gdshader) I'm noticing some rather 'unexpected' behavior with PROJECTION_MATRIX and depth_texture. I'm aware that this is part of the reverse z buffer breaking changes, but is there any way around this? Sorry to clutter this thread with a rather novice question, but I'd be very appreciative if someone could explain what exactly changes about these values and how to bodge something together to adapt existing shaders.

Khasehemwy commented:
> edit: managed to figure this out, leaving this for anyone else who's struggling
> 
> If you'd like to use Reverse-Z on existing user shaders, anywhere `INV_VIEW_MATRIX` is used needs to be inverted within the shader. Here's an example of how I managed this - might not be the best as I'm not a shader wizard, but it works.
> 
> ```
> mat4 modifiedMatrix = INV_VIEW_MATRIX;
> modifiedMatrix[3][0] = -modifiedMatrix[3][0];
> modifiedMatrix[3][1] = -modifiedMatrix[3][1];
> modifiedMatrix[3][2] = -modifiedMatrix[3][2];
> ```
> 
> You can then use `modifiedMatrix` as you would normally use `INV_VIEW_MATRIX`.

reversed-z only affects the projection matrix, should have nothing to do with the view matrix(?). The INV_PROJECTION_MATRIX will also auto change, shader developers not need to change it.
This PR will indeed break shader compatibility. Most shaders should only need to change the depth comparison method.

violetfrost commented:
Then it would appear that I've found an issue, unless I'm mistaken which I very well could be considering my novice state. I wrote a very crude test minimum-reproduction shader:

```
shader_type spatial;

varying mat4 modifiedMatrix;
const bool DoFix = true;
void vertex() {
	modifiedMatrix = INV_VIEW_MATRIX;
    modifiedMatrix[3][0] = -modifiedMatrix[3][0];
    modifiedMatrix[3][1] = -modifiedMatrix[3][1];
    modifiedMatrix[3][2] = -modifiedMatrix[3][2];
}

void fragment()
{
	if(DoFix)
	{
		ALBEDO = modifiedMatrix[3].xyz;
	} else {
		ALBEDO = INV_VIEW_MATRIX[3].xyz;
	}
}
```

with `DoFix` set to true, this produces the same result as in a normal current build of Godot. With it set to false, it produces the inverse. If I use this shader in a normal build of Godot, it does the opposite: DoFix reproduces the wonky matrix seen in the PR version. I don't *think* there's anything wrong on my end in my larger shader, but it's 1AM and I am rather sleep deprived so do with that what you will.

clayjohn commented:
@violetfrost Without seeing your shader, it is impossible to pinpoint what the issue is. This PR doesn't alter the `INV_VIEW_MATRIX`, so your messages are a little confusing. You shader code also doesn't read from the depth buffer, so it really doesn't seem to have anything to do with this PR. 

I have tested this PR with the following shader code (copied from the [docs](https://docs.godotengine.org/en/latest/tutorials/shaders/advanced_postprocessing.html)) and it works correctly:

```
uniform sampler2D depth_tex : hint_depth_texture;

void fragment() {
	float depth = texture(depth_tex, SCREEN_UV).x;
	vec3 ndc = vec3(SCREEN_UV * 2.0 - 1.0, depth);
	vec4 view = INV_PROJECTION_MATRIX * vec4(ndc, 1.0);
	view.xyz /= view.w;
	float linear_depth = -view.z;
	ALBEDO = vec3(linear_depth)/40.0;
}
```

Is there any chance you are applying `INV_VIEW_MATRIX` directly to the raw value from the depth (i.e. not converting to view space first)? 

As a reminder, you need to convert to view space using `INV_PROJECTION_MATRIX` before you convert _from_ view space into world space with `INV_VIEW_MATRIX`

clayjohn commented:
@Khasehemwy Thank you for merging my PR. After my last update I thought about this a little more and I have a feeling that post processing is broken. At least for SSAO, SSIL, and SSR. They all linearize at some point and I'm not sure that our current depth linearization function handles reversed z values properly. I'm going to investigate a bit more tomorrow or Monday to confirm. Ideally I would like to merge this PR early next week.

violetfrost commented:
> Is there any chance you are applying `INV_VIEW_MATRIX` directly to the raw value from the depth (i.e. not converting to view space first)?

No - and in the minimum reproduction shader I wrote INV_VIEW_MATRIX returns a completely different output (looks to be directly inverted) on both the latest PR from @clayjohn [a83eb16fb] and the latest official build 4.2 [b09f793f5]. [Note the value of DoFix in each respective editor window.](https://i.imgur.com/A8xaQtX.png)

clayjohn commented:
I can't reproduce your issue with your provided code. Testing this PR and testing 4.2.1. I get the same results. 

![Screenshot from 2024-03-09 23-21-07](https://github.com/godotengine/godot/assets/16521339/fb146395-ef9b-4848-8e46-5674fc8cce20)
![Screenshot from 2024-03-09 23-20-45](https://github.com/godotengine/godot/assets/16521339/b7f0a6f6-174d-48e2-a058-147b9572dbd8)

At this point, you should probably check a couple of things:
1. Ensure you are building from this PR and not another one by accident (i.e. is it possible you have pulled this PR on top of unrelated local changes?)
2. Ensure you have the same build settings in this branch and when building 4.2.1
3. Ensure that nothing else is affecting your camera matrix (i.e. changes to the editor camera, switching from editor camera to a Camera3D node, etc.)

I really want to highlight that this PR does not touch that matrix at all. So it is extremely unlikely that the issue you are seeing is related to this PR.

violetfrost commented:
A-ha! You were right - my sleep deprivation got the better of me. I was testing on the wrong branch - my apologies! Great work you all are doing :)

dsnopek commented:
> I think we should try to fix the WebXR regression after merging as it might take a bit for us to narrow down the issue.

Just to be clear, the issue isn't specific to WebXR - I had been testing with OpenXR on the Meta Quest. I think it probably affects stereo rendering with OpenGL on any XR platform.

Ansraer commented:
Thanks for testing this in VR David!

> Ideally I would like to merge this PR early next week.

What about the cubemap and omnilight shadows? I looked at the changes you made to the glsl files and didn't see you address that anywhere. According to GitHub, the only change to shadercode is a fix for shadow pancaking (to fix the bug mrjustaguy found).

clayjohn commented:
> Thanks for testing this in VR David!
> 
> > Ideally I would like to merge this PR early next week.
> 
> What about the cubemap and omnilight shadows? I looked at the changes you made to the glsl files and didn't see you address that anywhere. According to GitHub, the only change to shadercode is a fix for shadow pancaking (to fix the bug mrjustaguy found).

They are stored in linear space. They don't benefit from reversed z to the same extent. And I don't think it's possible to avoid reversing them manually. With omnilight shadows we aren't comparing the depth of the shadow map to the depth in some coordinate space. We are calculating the distance from the surface to the light in world space and then comparing that to the shadow map. This means there is no way to bake the calculation into a projection matrix. 

Additionally, shadows are stored in 16 bit textures. They don't benefit from reversed z anyway

Calinou reviewed (commented):
I've tested the latest revision of this PR rebased on top of `master` https://github.com/godotengine/godot/commit/22c20cea6e59a8a95585adb866c007a55c428806 on Linux + NVIDIA on https://github.com/Calinou/godot-rendering-tests. Here are all the images that differ significantly (due to broken effects in this PR):

Test | `master` https://github.com/godotengine/godot/commit/22c20cea6e59a8a95585adb866c007a55c428806 | This PR
-:|-|-
DoF Far | ![test_depth_of_field_far](https://github.com/godotengine/godot/assets/180032/546f9c43-ecd3-47e5-a59f-df738bb86a8a) | ![test_depth_of_field_far](https://github.com/godotengine/godot/assets/180032/902b3049-4f80-4944-ae0b-6a0a551ce00c)
SSIL | ![test_ssil](https://github.com/godotengine/godot/assets/180032/77745a61-9872-4799-b27a-9a56f3cc6a8b) | ![test_ssil](https://github.com/godotengine/godot/assets/180032/f13675e2-73f0-4fd9-8997-8a261a9eefef)
SSIL + SSAO | ![test_ssil_ssao](https://github.com/godotengine/godot/assets/180032/a6b3059c-eb2e-420c-a25b-cdf27f7747c7) | ![test_ssil_ssao](https://github.com/godotengine/godot/assets/180032/85723bdf-3f40-4c26-9f6d-e7a2cbf8842c)
DoF Near | ![test_depth_of_field_near](https://github.com/godotengine/godot/assets/180032/bd7c5980-c684-48f0-803f-25c4e8c30568) | ![test_depth_of_field_near](https://github.com/godotengine/godot/assets/180032/20898308-c336-4df6-a7cf-74b732a7205e)
SSR | ![test_ssr](https://github.com/godotengine/godot/assets/180032/9224db2f-4e3a-4beb-b206-8ac78913f507) | ![test_ssr](https://github.com/godotengine/godot/assets/180032/1a73f0d4-38b7-4dcb-8ccd-b286ca0078a3)
SDFGI | ![test_sdfgi](https://github.com/godotengine/godot/assets/180032/9e2b3194-e380-4d6c-b993-df3aa2def9df) | ![test_sdfgi](https://github.com/godotengine/godot/assets/180032/e54b26cc-cf1c-4107-8f65-d2e88c6496c2)
FogVolume | ![test_fog_volume](https://github.com/godotengine/godot/assets/180032/c60ae202-1107-4caa-90f8-bc628104d769) | ![test_fog_volume](https://github.com/godotengine/godot/assets/180032/eeef5e59-82b8-4fc5-a391-b9899d08b119)
SSAO | ![test_ssao](https://github.com/godotengine/godot/assets/180032/2f3eda4a-81f6-4c45-b7ef-d6f5459f837e) | ![test_ssao](https://github.com/godotengine/godot/assets/180032/d244eb2d-935b-47a2-b1f9-de23ac94b3db)
Volumetric Fog | ![test_volumetric_fog](https://github.com/godotengine/godot/assets/180032/3f0ac912-faf8-4d87-a8b2-c5d040b96c93) | ![test_volumetric_fog](https://github.com/godotengine/godot/assets/180032/c75cf377-d385-4c17-a612-b672fc7987d0)
VoxelGI | ![test_voxel_gi](https://github.com/godotengine/godot/assets/180032/b8059d2f-fa08-4ce5-acbd-900243bfba09) | ![test_voxel_gi](https://github.com/godotengine/godot/assets/180032/54eda2ec-38a0-4ecd-8d47-aeb297ec8335)

To check images that differ significantly, run the project once, copy the `results` folder to `results_master`, then switch to a build of this PR and run the project again. Install [dssim](https://github.com/kornelski/dssim) and run:

```bash
(for img in results/*/*.png; do dssim $img ${img//results/results_master}; done) | sort
```

Images that differ the most from `master` are displayed at the bottom of the list.

clayjohn commented:
Thanks @Calinou I'm working on fixes these cases now:

- [x] SSR
- [x] SSAO
- [x] SSIL
- [ ] DoF
- [ ] SDFGI
- [ ] VoxelGI
- [ ] VolumetricFog

clayjohn commented:
This should be ready to merge once https://github.com/Khasehemwy/godot/pull/2 is merged. I have now tested everything except XR. There are a few edge cases that might arise in XR so more testing is welcome, especially with GI and effects

Khasehemwy commented on servers/rendering/renderer_rd/shaders/effects/bokeh_dof_raster.glsl:
I noticed that there is a separate treatment for whether it is orthogonal, which may need to be changed to:
```glsl
if (params.orthogonal) {
	depth = -((depth - (params.z_far + params.z_near) / (params.z_far - params.z_near)) * (params.z_far - params.z_near)) / 2.0;
} 
```
Or simplify it (I think it’s better):
```glsl
if (params.orthogonal) {
	depth = -(depth * (params.z_far - params.z_near) - (params.z_far + params.z_near)) / 2.0;
} 
```

The same applies to other places. I am going to test it (but may be slower). You can also test it.

Khasehemwy replied:
There is indeed problems. After the change, it will be consistent with 4.2.

And I found a new bug. When the camera is in orthogonal mode, there are some problems with ssr and gi (godot4.2)

dsnopek commented:
Using the latest code on this PR, I've made a quick video showing the sky issue with XR and the compatibility renderer on a Meta Quest 3:

https://github.com/godotengine/godot/assets/191561/090084e6-a509-491f-8e5d-0f695cefeaf9

Rather than staying fixed in the right place, the horizon moves depending on how I turn my head (aka the camera). It sort of feels like it's moving in the inverse direction than it should in order for it to appear fixed in place.

And, it's actually worse than it appears in the video, because the left and right eye are showing something different (the video is only of the left eye): if I tilt my head to the left or right, the horizon will move one direction in the left eye, and the opposite direction in the right eye - breaking proper stereo perspective.

As I mentioned previously, this probably doesn't happen with the Vulkan Mobile renderer, only the OpenGL renderer.

Khasehemwy commented:
It may be Vulkan is incorrect after modification. 
I saw that sky.cpp has been modified:
![image](https://github.com/godotengine/godot/assets/57032017/113f1a57-75e3-4706-a463-3e8a82d62a57)
 You can try changing it to the following, and test Vulkan again:
![image](https://github.com/godotengine/godot/assets/57032017/00fec069-b860-4aee-a1be-f60bbe39d6f4)

I don’t know if it can be fixed. Sorry, I don't have an xr device to test it.

dsnopek commented:
@Khasehemwy Thanks, that fixed the issue on the Vulkan Mobile renderer!

How can we translate that fix to OpenGL as well?

I've attached the changes I made per your screenshot as a patch, just so it's clear to everyone exactly what the changes were.

[xr-vulkan-mobile-sky-fix.patch.txt](https://github.com/godotengine/godot/files/14603233/xr-vulkan-mobile-sky-fix.patch.txt)

Khasehemwy commented:
> @Khasehemwy Thanks, that fixed the issue on the Vulkan Mobile renderer!

@dsnopek Thank you, Nice! I commit a change, maybe can fix OpenGL issue.

dsnopek commented:
Thanks! The lastest changes do fix the sky issue on OpenGL, but create a new issue - it seems like it's un-reversing the depth buffer?

Here's a screenshot with Vulkan Mobile:

![org godotengine openxr vendors demo-20240314-094504](https://github.com/godotengine/godot/assets/191561/7e60ba32-9857-439b-9b71-6860214f1f62)

And here's with OpenGL:

![org godotengine openxr vendors demo-20240314-094408](https://github.com/godotengine/godot/assets/191561/55d633ff-39ff-4d29-b0a9-d3e025714acf)

clayjohn commented:
@dsnopek I'm a little bit skeptical of `p_render_data->view_projection[v]`. It seems like it is combining a view-space operation into the projection:

https://github.com/godotengine/godot/blob/da945ce6266ce27ba63b6b08dc0eb2414594f7cb/servers/rendering/renderer_scene_render.cpp#L184-L185

BastiaanOlij commented:
> @dsnopek I'm a little bit skeptical of `p_render_data->view_projection[v]`. It seems like it is combining a view-space operation into the projection:
> 
> https://github.com/godotengine/godot/blob/da945ce6266ce27ba63b6b08dc0eb2414594f7cb/servers/rendering/renderer_scene_render.cpp#L184-L185
> 
> Okay, ya it is. It a little per-eye correction in view space. It looks like we encode a transform that goes from view space to world space, back to view space then into clip space. The reverse-z portion should only apply to the clip space portion. But we don't have that at the rendering driver level! I think we need to keep the view-space operations out of the projection and then just combine them at the last minute. If `view_projection` is accessed anywhere else it is probably messing up a lot of things as a result

Ah this makes lots of sense Clay. This is the thing Juan came up with so we don't need to add a lot of extra shader code to handle the eyes separately in all sorts of processes.

The problem is that it's not just the eye offset left/right, but we can also end up moving the center point backwards so the combined frustum is usable. That shift now suddenly gets reversed as well with the reverse-z causing the weird wobble. It'll just be far more pronounced on headsets like the Quest. 

This would impact Vulkan as well and is not limited to OpenGL

BastiaanOlij commented:
Actually let me do some more testing because the more I think of this, the more this should still work, there must be a shortcut we're taking in the unproject on sky.

BastiaanOlij commented:
Ok, I fixed it all up, we were applying patches over patches to try and work around the real issue. There were two:

1. We have to unproject at the far plane, not the near plane due to stereo having asymmetrical view frustums. If you unproject at the near plane (which we now do because we flipped near and far around) you're basically getting your sky as if the sky is inside of your eyeballs, hence the really weird shifts.
2. Once I fixed 1, GLES worked perfectly, Vulkan suddenly broke. Turns out we were creating a GLES projection matrix in the Vulkan renderer.

https://github.com/Khasehemwy/godot/pull/3 fixes these issues and undoes some of the workaround patches that were wrong.

I also found out that we're being really messing with our y-flip. Sky is actually doing the y-flip in the shader, and we're countering it in places by y-flipping when we don't need y-flipping and visa versa. It's a miracle this is working correctly in reflection probes, but it is:).
One day we should really clean this up and make everything render the same way up.

The volumetric code is a bit of a hack. What we're doing here is un-projecting the stereo view and reprojecting to mono. That's fundamentally flawed. We really need to do a stereo version of our volumetric buffer. That's a whole different thing for another day.

dsnopek commented:
@BastiaanOlij Thanks so much!

I've retested the latest version of this PR (which includes Bastiaan's changes) using both OpenGL and the Vulkan mobile renderer on a Meta Quest Pro, and everything (including sky rendering) seems to be working correctly now. :-)

Khasehemwy commented:
Looks like this PR is in good shape now. Is there anything else I can do to assist with the merge?

clayjohn reviewed (approved):
Looks good to me now (pending squash and rebase). I have tested extensively on desktop devices and I trust the testing of Bastiaan and David on VR devices. 

Once this is squashed, we can go ahead and merge it for 4.3

lyuma commented:
Sorry if this is an obvious question but this is a pretty big breaking change (as noted in the OP) that will impact user shaders.
How can I as a user modify my shaders to be compatible with both the 4.2 and 4.3 rendering design?

I am surprised that a change like this is better in all cases and would have no downsides. If this is true, do all other major engines also use this approach on dx/vulkan? (Should this perhaps be a project setting?)

I saw XR mentioned. Are there any interoperability concerns related to submitting reversed depth buffer to OpenXR?

Ansraer commented:
> Sorry if this is an obvious question but this is a pretty big breaking change (as noted in the OP) that will impact user shaders.
How can I as a user modify my shaders to be compatible with both the 4.2 and 4.3 rendering design?

Most of your shaders should be completely fine, this only has an impact on you if you either manually write to depth or read it. That's why we really want to get this merged before the compositor PR makes writing custom post processing shaders much easier.

Should you run into any problems all you need to know is that the depth buffer now goes from `1.0-0.0` instead of `0.0-1.0`. This means that you might have to invert any compare ops you do and maybe switch additions and subtraction operations.

> I am surprised that a change like this is better in all cases and would have no downsides. If this is true, do all other major engines also use this approach on dx/vulkan? (Should this perhaps be a project setting?)

Yes, this is the approach that is used by any recent-ish engine. It has major advantages when it comes to precision and accuracy with 0 downsides, so there is no point in adding a project setting to turn it off (as a matter of fact that would actually be counterproductive, since custom post-processing assets could break depending on the state of that project setting).
The only reason why this wasn't a thing in godot already is because this optimization has no visible advantage in opengl, and people forgot to add it when the engine moved to vulkan.

> I saw XR mentioned. Are there any interoperability concerns related to submitting reversed depth buffer to OpenXR?

Not that I am aware of. As said before, rev depth is the current state of the art and recommended by every hardware vendor out there. I have little experience with developing for VR, but I would be seriously surprised if rev depth was a problem.

darksylinc commented:
> I am surprised that a change like this is better in all cases and would have no downsides.

Reverse Z has the following conditions:

_Much_ Better in **all** scenarios if:

1. Depth Buffer is set to `D32_FLOAT`. Only very old GPUs (more than 15+ years old) do not support it.
     - However on Linux + X11 + OpenGL (i.e. Compatibility), Mesa does not support `D32_FLOAT` for the render window. Fortunately Godot does not render directly to the render window so this not an issue but it can be for other engines.
2. In OpenGL (i.e. compatibility) glClipControl is supported and is set to range [0; 1]. This is not a given for OpenGL ES (i.e. mobile).

_Slightly_ worse if:

1. Depth Buffer is set to `D24_UNORM` (it should not be an issue in Godot) or
2. glClipControl is not supported (i.e. Compatibility in Mobile).

> If this is true, do all other major engines also use this approach on dx/vulkan? (Should this perhaps be a project setting?)

Other engines chose to either move to reverse Z as well, or allow for optionally support both.

Supporting both options requires a lot of testing and maintenance which is why the Godot team chose not to support the old regular Z at all. Often engines that support both options end up having bugs when using regular Z because very few people actually use that code path (due to the massive superiority of reverse Z).

>  How can I as a user modify my shaders to be compatible with both the 4.2 and 4.3 rendering design?

Godot devs will have to answer you this one. You do make a point that perhaps Godot 4.3 should define "something" to signal user shaders that reverse Z is being used; in case a user wants to support both versions (<= 4.2 + >= 4.3).

lyuma commented:
Ok your answers all make sense. I am pretty sold.

the main thing then is having some docs on porting shaders. For example one shader here is hardcoded in the godot engine source and probably needs to be pretty:
https://github.com/godotengine/godot/blob/master/editor/plugins/skeleton_3d_editor_plugin.cpp#L919..L920

In terms of user shaders, godot vrm does something with the projection space z coordinate for outlines:
https://github.com/V-Sekai/godot-vrm/blob/d88d2cc5a629d21a33196ef33a71f4bcb2736916/addons/Godot-MToon-Shader/mtoon_common.gdshaderinc#L103
People download this code and expect it to work in both 4.2 and 4.3 so I might need to come up with some heuristic to check for the inverted z buffer... I assume it should be possible to determine from the sign of some element of the projection matrix?

BastiaanOlij commented:
> I have little experience with developing for VR, but I would be seriously surprised if rev depth was a problem.

I did a little research on this and sadly, non of the APIs in OpenXR that allow us to submit a depth buffer, provide the ability to tell the XR runtime we're storing reverse-Z, meaning they will be incorrectly interpreted.

Right now we only use the composition layer depth extension, and that is generally turned off. This discussion actually made me find a bug in that logic that might explain some issues we've had in the past.

Metas Application Spacewarp will also be effected by this.

That said, seeing many game engines are switching over to this, it's something that I have raised within the working group, hopefully that will lead to discussions to eventually add support for this.

Khasehemwy commented:
@clayjohn Awesome! Thank you all! All the submissions have been squashed now.

Khasehemwy commented:
> I assume it should be possible to determine from the sign of some element of the projection matrix?

@lyuma Yes, you can determine whether it's reversed-z in the shader by checking the value of the `PROJECTIONMATRIX`. When reversed-z is applied, `columns[2][2] > 0` `( columns[2][2] = (p_z_far + p_z_near) / (p_z_far - p_z_near) )`. The same principle applies to orthogonal as well ( `columns[2][2] > 0` `( columns[2][2] = 2.0 / (p_zfar - p_znear); )` ). Of course, we should assume that `p_z_far > p_z_near` by default.

AThousandShips commented on servers/rendering/renderer_rd/environment/sky.cpp:
```diff
@@ -1184,10 +1185,12 @@ void SkyRD::setup_sky(RID p_env, Ref<RenderSceneBuffersRD> p_render_buffers, con
 	for (uint32_t i = 0; i < p_view_count; i++) {
 		Projection view_inv_projection = (correction * p_view_projections[i]).inverse();
 		if (p_view_count > 1) {
+			// Reprojection is used when we need to have things in combined space,
```
```suggestion
			// Reprojection is used when we need to have things in combined space.
```

AThousandShips commented on drivers/gles3/shaders/sky.glsl:
```diff
@@ -139,9 +139,11 @@ void main() {
 	vec4 unprojected = multiview_data.inv_projection_matrix_view[ViewIndex] * unproject;
 	cube_normal = unprojected.xyz / unprojected.w;
+
+	// Unproject will give us the position between the eyes, need to re-offset
```
```suggestion
	// Unproject will give us the position between the eyes, need to re-offset.
```

clayjohn commented:
> Ok your answers all make sense. I am pretty sold.
> 
> the main thing then is having some docs on porting shaders. For example one shader here is hardcoded in the godot engine source and probably needs to be edited: https://github.com/godotengine/godot/blob/master/editor/plugins/skeleton_3d_editor_plugin.cpp#L919..L920
> 
> In terms of user shaders, godot vrm does something with the projection space z coordinate for outlines: https://github.com/V-Sekai/godot-vrm/blob/d88d2cc5a629d21a33196ef33a71f4bcb2736916/addons/Godot-MToon-Shader/mtoon_common.gdshaderinc#L103 People download this code and expect it to work in both 4.2 and 4.3 so I might need to come up with some heuristic to check for the inverted z buffer... I assume it should be possible to determine from the sign of some element of the projection matrix?

@lyuma Neither of the shaders you linked will need to be modified. The only user shaders that will need to be modified are those that adjust `DEPTH` or read from the depth_texture without using `PROJECTION_MATRIX` or `INV_PROJECTION_MATRIX`. In practice, it almost never makes sense to manipulate depth in clip space, so I don't think we will run into this issue often. 

In other words:
```
// Compatibility is broken (need to flip the sign)
float depth = texture(depth_tex, SCREEN_UV).r;
vec4 clip = PROJECTION_MATRIX * vec4(VERTEX, 1.0);
clip.xyz /= clip.w;
if (depth < clip.z) {
    ...
}

// This is fine
float depth = texture(depth_tex, SCREEN_UV).r;
vec3 ndc = vec3(SCREEN_UV * 2.0 - 1.0, depth);
vec4 view = INV_PROJECTION_MATRIX * vec4(ndc, 1.0);
view.xyz /= view.w;
float linear_depth = -view.z;
if (view < VERTEX.z) {
    ...
}
```

Khasehemwy commented:
Sorry @AThousandShips , I clicked on the wrong one and got a review😟

akien-mga commented:
I amended the commit message to clean the description which still included past fixup commit titles, and retitled it to be clearer, especially regarding the fact that it's changing the convention, not adding a new option.

akien-mga commented:
Thanks! And congrats for your first – but massive! – merged Godot contribution :tada:

clayjohn commented:
Adding a note for the dev blog post:

> This dev release switches our depth buffer to using a "reverse z" projection. This means that instead of the depth going from 0 (close to viewer) to 1 (far from view), it goes from 1 to 0. This is a enhancement found in most modern engines and it significantly improves depth precision by taking advantage of how floating point numbers are stored. This is a breaking change, however, we think the compatibility breakage is minimal and well worth the benefits. 
> 
> In particular, any operation that uses clip space values directly from the depth buffer are now reversed. That includes uses of the depth buffer in custom render hook callbacks and uses of the `depth_texture` hint in shaders. That being said, we are confident that most users are reading depth to reconstruct view space, or world space depth which still works the same as before. For example the below is a very common pattern that will continue to work without modification:
```
float depth = texture(depth_tex, SCREEN_UV).r;
vec3 ndc = vec3(SCREEN_UV * 2.0 - 1.0, depth);
vec4 view = INV_PROJECTION_MATRIX * vec4(ndc, 1.0);
view.xyz /= view.w;
float linear_depth = -view.z;
if (view < VERTEX.z) {
    ...
}
```
> As long as you are using the `INV_PROJECTION_MATRIX` to convert the depth into view space, no modifications are necessary. If you are working in clip space directly, you will need to reverse the sign of depth comparisons manually. For example take the following code:
```
float depth = texture(depth_tex, SCREEN_UV).r;
vec4 clip = PROJECTION_MATRIX * vec4(VERTEX, 1.0);
clip.xyz /= clip.w;
if (depth < clip.z) {
    ...
}
```
> To maintain compatibility, you will have to change the comparison from `depth < clip.z` to `depth > clip.z`. The same is true for any code that is directly working with a depth value, or a clip space value returned by using `PROJECTION_MATRIX`. 
> 
> Overall, we regret to break compatibility in this way, but we felt strongly that the tradeoff was worthwhile and the alternative was to wait until Godot 5 which will be many years away.
